File: src/breadth_first_search/BFS_adjmat.java
package breadth_first_search;

import java.util.LinkedList;
import java.util.Queue;

public class BFS_adjmat {

	static class BFSTraversal {
		int vertex;
		boolean visited[];
		int mat[][];

		public BFSTraversal(int v) {
			vertex = v;
			mat = new int[vertex][vertex];
			visited = new boolean[vertex];
		}

		void addEdge(int u, int v) {
			mat[u][v] = 1;
		}

		void BFS(int s) {
			Queue<Integer> q = new LinkedList<>();
			q.add(s);

			while (q.size() > 0) {
				int a = q.peek();
				System.out.println(a);
				visited[a] = true;
				q.poll();
				for (int i = 0; i < vertex; i++) {
					if (mat[a][i] == 1 && !visited[i]) {
						q.add(i);
					}
				}
			}
		}
	}

	public static void main(String[] args) {
		int v = 6;
		BFSTraversal bfst = new BFSTraversal(v);
		bfst.addEdge(0, 1);
		bfst.addEdge(0, 2);
		bfst.addEdge(1, 3);
		bfst.addEdge(2, 4);
		bfst.addEdge(2, 5);
		
		bfst.BFS(0);
	}

}

File: src/breadth_first_search/BFS_adjmat.go
package main

import "fmt"

var mat [10][10]int

func size(queue []int) int {
	return len(queue)
}

func enqueue(queue []int, element int) []int {
	queue = append(queue, element)
	return queue
}

func peek(queue []int) int {
	return queue[0]
}

func dequeue(queue []int) []int {
	return queue[1:]
}

func addEdge(u int, v int) {
	mat[u][v] = 1
}

func bfs(s int, visited [5]int) {
	var queue []int
	queue = enqueue(queue, s)

	for size(queue) > 0 {

		var a int
		a = peek(queue)
		fmt.Println(a)
		visited[a] = 1

		queue = dequeue(queue)

		for i := 0; i < 5; i++ {
			if mat[a][i] == 1 && visited[i] == 0 {
				queue = enqueue(queue, i)
			}
		}
	}
}

func main() {
	var visited [5]int

	addEdge(0, 1)
	addEdge(0, 2)
	addEdge(1, 3)
	addEdge(1, 4)

	bfs(0, visited)
}
